tik/lib/CartographerKernel.cpp

```diff
@@ -952,166 +1310,346 @@ namespace TraceAtlas::tik
+            BasicBlock *tmp = BasicBlock::Create(TikModule->getContext(), "", KernelFunction);
+            IRBuilder<> builder(tmp);
+            builder.CreateBr(Exit);
+            exitMap[exit] = tmp;
+        }
+
         for (auto fi = KernelFunction->begin(); fi != KernelFunction->end(); fi++)
         {
-            auto BB = cast<BasicBlock>(fi);
-            for (BasicBlock::iterator bi = BB->begin(); bi != BB->end(); bi++)
+            auto block = cast<BasicBlock>(fi);
+            auto term = block->getTerminator();
+            if (term != nullptr)
             {
-                auto *inst = cast<Instruction>(bi);
-                for (unsigned int arg = 0; arg < inst->getNumOperands(); arg++)
+                for (uint32_t i = 0; i < term->getNumSuccessors(); i++)
                 {
-                    Value *inputOp = inst->getOperand(arg);
-                    if (auto op = dyn_cast<Operator>(inputOp))
+                    auto suc = term->getSuccessor(i);
+                    if (suc->getParent() != KernelFunction)
                     {
-                        RemapOperands(op, inst, VMap);
+                        //we have an exit
+                        term->setSuccessor(i, exitMap[suc]);
                     }
                 }
-                RemapInstruction(inst, VMap, llvm::RF_None);
             }
         }
+
+        auto phi = exitBuilder.CreatePHI(Type::getInt8Ty(TikModule->getContext()), (uint32_t)Exits.size());
+        for (const auto &exit : Exits)
+        {
+            if (exitMap.find(IDToBlock[exit->Block]) == exitMap.end())
+            {
+                throw AtlasException("Block not found in Exit Map!");
+            }
+            phi->addIncoming(ConstantInt::get(Type::getInt8Ty(TikModule->getContext()), (uint64_t)exit->Index), exitMap[IDToBlock[exit->Block]]);
+        }
+        exitBuilder.CreateRet(phi);
+
+        IRBuilder<> exceptionBuilder(Exception);
+        exceptionBuilder.CreateRet(ConstantInt::get(Type::getInt8Ty(TikModule->getContext()), (uint64_t)IDState::Artificial));
     }
 
     void CartographerKernel::FixInvokes()
     {
-        auto F = TikModule->getOrInsertFunction("__gxx_personality_v0", Type::getInt32Ty(TikModule->getContext()));
-        for (auto fi = KernelFunction->begin(); fi != KernelFunction->end(); fi++)
+        TikModule->getOrInsertFunction("__gxx_personality_v0", Type::getInt32Ty(TikModule->getContext()));
+        for (auto &fi : *KernelFunction)
         {
-            for (auto bi = fi->begin(); bi != fi->end(); bi++)
+            for (auto bi = fi.begin(); bi != fi.end(); bi++)
             {
                 if (auto ii = dyn_cast<InvokeInst>(bi))
                 {
                     auto a = ii->getLandingPadInst();
                     if (isa<BranchInst>(a))
                     {
-                        auto unwind = ii->getUnwindDest();
+                        throw AtlasException("Exception handling is not supported");
```

Comment:
This almost certainly removes support for all invoke instructions.

Reply:
it does and thats by design. When this code is exercised and results in the generation of invoke instructions, there is a module error, either because the personality function is from another module, or because the invoke instruction uses a null type, which is not allowed

Reply:
The personality function is provided by libcpp. It shouldn't be defined in the other module. I would really like us to try and fix this.

Reply:
not what i've observed. a good example of this is opencv/contours2. we should work through it later.